### PR TITLE
Allow ignored columns for mview to be specified at the subscription level

### LINKED ISSUE #30243: [Issue] Allow ignored columns for mview to be specified at the subscription level
This issue is automatically created based on existing pull request: magento/magento2#29692: Allow ignored columns for mview to be specified at the subscription level

---------

### Summary (*)
Magento 2.4-develop
Current configuration options for indexer subscriptions are limited - only allowing a table to be specified, rather than specific columns that we may be interested in.
Likewise `\Magento\Framework\Mview\View\Subscription` has the argument `$ignoredUpdateColumns`, which by default only includes `updated_at`. This argument is used to ignore columns for update triggers globally.

This change adds another argument to `\Magento\Framework\Mview\View\Subscription` to allow for columns to be ignored for a specific view/table combination (i.e. at the subscription level).

**Example use-case**:
You may wish to subscribe to the `salesrule` table for a custom indexer (e.g. view id of "my_custom_view"), but with current functionality, this results in a reindex occurring any time an order is placed which had the rule applied, due to the `times_used` column being updated.

With the following `di.xml` entries, we can specifically ignore the `times_used` column for our indexer view:
```
<type name="Magento\Framework\Mview\View\Subscription">
    <arguments>
        <argument name="ignoredUpdateColumnsBySubscription" xsi:type="array">
            <item name="my_custom_view" xsi:type="array">
                <item name="salesrule" xsi:type="array">
                    <item name="times_used" xsi:type="string">times_used</item>
                </item>
            </item>
        </argument>
    </arguments>
</type>
```

### Examples (*)

1. Add `di.xml` configuration for an existing view - e.g. `cataloginventory_stock`:
```
<type name="Magento\Framework\Mview\View\Subscription">
    <arguments>
        <argument name="ignoredUpdateColumnsBySubscription" xsi:type="array">
            <item name="cataloginventory_stock" xsi:type="array">
                <item name="cataloginventory_stock_item" xsi:type="array">
                    <item name="low_stock_date" xsi:type="string">low_stock_date</item>
                </item>
            </item>
        </argument>
    </arguments>
</type>
```
2. Set all indexers to update on save (this will remove existing triggers)
3. Set all indexers to update by schedule (this will create triggers again)
4. Observe that updating `cataloginventory_stock_item.low_stock_date` does not create a record in `cataloginventory_stock_cl`, but does create records in `catalogsearch_fulltext_cl` and `catalog_product_price_cl`

### Proposed solution

### PR DESCRIPTION
**Pull request by aligent-lturner**
### Description (*)
Current configuration options for indexer subscriptions are limited - only allowing a table to be specified, rather than specific columns that we may be interested in.
Likewise `\Magento\Framework\Mview\View\Subscription` has the argument `$ignoredUpdateColumns`, which by default only includes `updated_at`. This argument is used to ignore columns for update triggers globally.

This change adds another argument to `\Magento\Framework\Mview\View\Subscription` to allow for columns to be ignored for a specific view/table combination (i.e. at the subscription level).

**Example use-case**:
You may wish to subscribe to the `salesrule` table for a custom indexer (e.g. view id of "my_custom_view"), but with current functionality, this results in a reindex occurring any time an order is placed which had the rule applied, due to the `times_used` column being updated.

With the following `di.xml` entries, we can specifically ignore the `times_used` column for our indexer view:
```
<type name="Magento\Framework\Mview\View\Subscription">
    <arguments>
        <argument name="ignoredUpdateColumnsBySubscription" xsi:type="array">
            <item name="my_custom_view" xsi:type="array">
                <item name="salesrule" xsi:type="array">
                    <item name="times_used" xsi:type="boolean">true</item>
                </item>
            </item>
        </argument>
    </arguments>
</type>
```

### Manual testing scenarios (*)

1. Add `di.xml` configuration for an existing view - e.g. `cataloginventory_stock`:
`app/code/Magento/CatalogInventory/etc/di.xml`
```
<type name="Magento\Framework\Mview\View\Subscription">
    <arguments>
        <argument name="ignoredUpdateColumnsBySubscription" xsi:type="array">
            <item name="cataloginventory_stock" xsi:type="array">
                <item name="cataloginventory_stock_item" xsi:type="array">
                    <item name="low_stock_date" xsi:type="boolean">true</item>
                </item>
            </item>
        </argument>
    </arguments>
</type>
```
2. Set all indexers to update on save (this will remove existing triggers)
3. Set all indexers to update by schedule (this will create triggers again)
4. Observe that updating `cataloginventory_stock_item.low_stock_date` does not create a record in `cataloginventory_stock_cl`, but does create records in `catalogsearch_fulltext_cl` and `catalog_product_price_cl`

### Questions or comments

Not sure if there's any tests that can be updated/added for this change. Please feel free to suggest them (or even add them!)

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#30243: Allow ignored columns for mview to be specified at the subscription level

### CODE REVIEW

File: lib/internal/Magento/Framework/Mview/View/Subscription.php [hunk 1]
```diff
@@ -209,7 +218,9 @@ protected function buildStatement($event, $changelog)
                     $describe = $this->connection->describeTable($tableName)
                 ) {
                     $columnNames = array_column($describe, 'COLUMN_NAME');
-                    $columnNames = array_diff($columnNames, $this->ignoredUpdateColumns);
+                    $ignoredColumns = array_merge($this->ignoredUpdateColumns,
```
Comment (BarnyShergold) [inline lib/internal/Magento/Framework/Mview/View/Subscription.php]: Formatting incorrect. If multi-line then all parameters must be on a line of their own - this is causing tests to fail.

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
Correct formatting to fix test & fix functional tests

File: lib/internal/Magento/Framework/Mview/View/Subscription.php (hunk 1 above)
Comment (aligent-lturner) [reply lib/internal/Magento/Framework/Mview/View/Subscription.php]: Have fixed in latest commit

**Comment by aligent-lturner** [comment]
@magento run all tests

**Comment by aligent-lturner** [comment]
@magento run all tests

**Review by BarnyShergold — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @BarnyShergold, thank you for the review.
[ENGCOM-8063](https://jira.corp.magento.com/browse/ENGCOM-8063) has been created to process this Pull Request
:eight_spoked_asterisk: @BarnyShergold, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by sidolov** [comment]
@magento create issue

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked according [Manual testing scenarios](https://github.com/magento/magento2/pull/29692#issue-470682497)
**Result:**
:x: **Before:**
Column update cannot be ignored
#26692Main

:heavy_check_mark: **After:**
Column update is ignored according configuration in `di.xml`
#29692PR

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @aligent-lturner .  Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. 
Could you cover your changes by integration test?

**Comment by aligent-lturner** [comment]
@engcom-Delta I'm not really sure how you would go about writing an integration test for this change. There's no existing integration or unit test around the existing `ignoredUpdateColumns` parameter. I'm more than happy for someone to suggest how such a test would be structured, but I'm not sure how simple it would be to test automatically, since it would effectively require parsing of generated trigger code. The change has passed the manual testing as shown above.

**Comment by engcom-Charlie** [comment]
@aligent-lturner I'll take care of test coverage.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by dhorytskyi** [comment]
Hi @aligent-lturner , did you consider extending mview.xml structure for this purpose? It can be like this:
```
...
<view id="cataloginventory_stock" class="..." group="indexer">
    <subscriptions>
        <table name="cataloginventory_stock_item" entity_column="product_id" ignore_columns="low_stock_date" />
        ...
    </subscriptions>
</view>
...
```
It will allow to keep all information about indexer configuration in one place. Multiple columns can be specified in a comma-separated way. In most cases, columns for excluding should be specified in the module which declares indexer, but it still can be changed through another module.

I can provide an implementation for this proposal.

**Comment by aligent-lturner** [comment]
@dhorytskyi I'll admit I didn't consider that, though I must say I'm not really keen on the idea of having comma-separated values in XML. I'm happy for you to provide such an implementation though.

**Comment by engcom-Charlie** [comment]
image

@aligent-lturner can you please implement `di.xml` configuration like this?
In this way, we will have the opportunity to change `ignore/not ignore` column through another module.
Also, could I ask you to update the unit test accordingly to this change?
Thank you.

**Comment by aligent-lturner** [comment]
@engcom-Charlie I have made the changes as requested.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by gabrieldagama — APPROVED** [review]
Hi, @aligent-lturner thanks for contributing. The changes look good.

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8063](https://jira.corp.magento.com/browse/ENGCOM-8063) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Result is same as in [comment](https://github.com/magento/magento2/pull/29692#issuecomment-702929969)

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Review by BarnyShergold — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @BarnyShergold, thank you for the review.
[ENGCOM-8063](https://jira.corp.magento.com/browse/ENGCOM-8063) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
**Result:**
:heavy_check_mark: Updating `cataloginventory_stock_item.low_stock_date` does not create a record in `cataloginventory_stock_cl` but does create records in `catalogsearch_fulltext_cl` and c`atalog_product_price_cl`
Peek 2020-11-18 12-23

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Comment by skovalenk** [comment]
Guys, can you please be aligned to best practices. There is specific configuration for Mview, it is called mview.xml. Why you did this in DI.xml. For what purpose?

**Comment by aligent-lturner** [comment]
> Guys, can you please be aligned to best practices. There is specific configuration for Mview, it is called mview.xml. Why you did this in DI.xml. For what purpose?

Because there is existing configuration for ignoring columns in di.xml for the class in question. It makes sense to keep it together there. I agree that ideally, the configuration would be in mview.xml, but this would also require updating the associated xsd and would then also separate the configuration for globally ignored columns and those ignored for a specific view/table

**Pull request merged**